### ISSUE TITLE
Tried to fix pretrained weights but performance still lower than reported

### ISSUE THREAD

**Issue by TomHeaven** [TAMU-VITA/ABD-Net#21]
I tried to replicate the results of the project on Market1501, DukeMTMC and MSMT17 using Pytorch 1.3.1.

First, I found some mismatches of duke and msmt17 pretrained weights with model definitions. The following code shoud be inserted
```python
for k, v in pretrain_dict.items():
            # fix pretrained weights for duke and msmt17
            if (args.load_weights.startswith('duke') or args.load_weights.startswith('msmt17')) and (k.startswith('branches.1.1.cam_module') or k.startswith('branches.1.1.pam_module')):
                k = k.replace('branches.1', 'branches.0')
            ######################
            if not (k in model_dict and model_dict[k].size() == v.size()):
                if k in model_dict:
                    print('param miss match', k, v.size(), model_dict[k].size())
                else:
                    print('extra key', k)
            else:
                print('matched key', k)
```
before the line
```python
pretrain_dict = {k: v for k, v in pretrain_dict.items() if k in model_dict and model_dict[k].size() == v.size()}
```
in file `train.py` to fix the problem.

Then, I evaluated pretrained weights using the following commands:
```
python3 train.py -s market1501 -t market1501 --evaluate --load-weights market_checkpoint_best.pth.tar --height 384  --use-of --abd-dan cam pam --abd-np 3 --shallow-cam --use-ow
python3 train.py -s dukemtmcreid -t dukemtmcreid --evaluate --load-weights duke_checkpoint_best.pth.tar --height 384  --use-of --abd-dan cam pam --abd-np 3 --shallow-cam --use-ow
python3 train.py -s msmt17 -t msmt17 --evaluate --load-weights msmt17_final_best.pth.zip.tar --height 384 --use-of --abd-dan cam pam --abd-np 3 --shallow-cam --use-ow
```
For flip evaluation, I added an aditional parameter `--filp-eval`
Here is the result:

Datasets | mAP | R1 | mAP(flip_eval) | R1(flip_eval) | mAP(paper) | R1(paper)
-- | -- | -- | -- | -- | -- | --
market1501 | 86.1 | 94.6 | 86.8 | 94.7 | 88.3 | 95.6
dukemtmc | 76.3 | 87.4 | 77.1 | 87.8 | 78.6 | 89.0
msmt17 | 48.4 | 73.8 | Insufficient Memory?  |  Insufficient Memory?   | 60.8 | 82.3

For makert1501 and dukemtmc, the replicated performance is a bit slower than reported. For msmt17, however, the performance gap is significant. 

Could the authors help me find the cause of performance drop? It is crucial since the paper reported performance gain compared with other competitive methods (such as IANet, DGNet) by a great margin (at least 8% in mAP and 5% in R1) on MSMT17. Personally, I think it's strange that this great advantage is not present on DukeMTMC-reid and Market1501 but present on MSMT17 only.

I noticed the original dataset page of msmt17 is not available now. What I got is with three folders only:
- MSMT17
-- bounding_box_train
-- bounding_box_test
-- query
The total number of gallery images is 82161 and total number of query images is 11659, which are correct. Dataset file msmt17.py is updated as 
``` python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import os
import glob
import re
import sys
import urllib
import tarfile
import zipfile
import os.path as osp
from scipy.io import loadmat
import numpy as np
import h5py
#from scipy.misc import imsave

from .bases import BaseImageDataset


class MSMT17(BaseImageDataset):
    """
    MSMT17

    Reference:
    Wei et al. Person Transfer GAN to Bridge Domain Gap for Person Re-Identification. CVPR 2018.

    URL: http://www.pkuvmc.com/publications/msmt17.html
    
    Dataset statistics:
    # identities: 4101
    # images: 32621 (train) + 11659 (query) + 82161 (gallery)
    # cameras: 15
    """
    #dataset_dir = 'msmt17'
    dataset_dir = 'MSMT17'

    def __init__(self, root='data', verbose=True, **kwargs):
        super(MSMT17, self).__init__()
        self.dataset_dir = osp.join(root, self.dataset_dir)
        self.train_dir = osp.join(self.dataset_dir, 'bounding_box_train')
        self.test_dir = osp.join(self.dataset_dir, 'bounding_box_test')
        self.query_dir = osp.join(self.dataset_dir, 'query')
        #self.list_train_path = osp.join(self.dataset_dir, 'MSMT17_V1/list_train.txt')
        #self.list_val_path = osp.join(self.dataset_dir, 'MSMT17_V1/list_val.txt')
        #self.list_query_path = osp.join(self.dataset_dir, 'MSMT17_V1/list_query.txt')
        #self.list_gallery_path = osp.join(self.dataset_dir, 'MSMT17_V1/list_gallery.txt')

        self._check_before_run()
        train = self._process_dir(self.train_dir)
        #val = self._process_dir(self.train_dir, self.list_val_path)
        query = self._process_dir(self.query_dir)
        gallery = self._process_dir(self.test_dir)

        #train += val
        #num_train_imgs += num_val_imgs

        if verbose:
            print("=> MSMT17 loaded")
            self.print_dataset_statistics(train, query, gallery)

        self.train = train
        self.query = query
        self.gallery = gallery

        self.num_train_pids, self.num_train_imgs, self.num_train_cams = self.get_imagedata_info(self.train)
        self.num_query_pids, self.num_query_imgs, self.num_query_cams = self.get_imagedata_info(self.query)
        self.num_gallery_pids, self.num_gallery_imgs, self.num_gallery_cams = self.get_imagedata_info(self.gallery)

    def _check_before_run(self):
        """Check if all files are available before going deeper"""
        if not osp.exists(self.dataset_dir):
            raise RuntimeError("'{}' is not available".format(self.dataset_dir))
        if not osp.exists(self.train_dir):
            raise RuntimeError("'{}' is not available".format(self.train_dir))
        if not osp.exists(self.test_dir):
            raise RuntimeError("'{}' is not available".format(self.test_dir))

    def _process_dir(self, dir_path):
        #with open(list_path, 'r') as txt:
        #    lines = txt.readlines()
        lines = [f for f in os.listdir(dir_path) if f.endswith('.jpg') ]
        dataset = []
        pid_container = set()
        for img_idx, img_name in enumerate(lines):
            pid, camid, _ = img_name.split('_')
            pid = int(pid) # no need to relabel
            camid = int(camid[1:]) - 1 # index starts from 0
            img_path = osp.join(dir_path, img_name)
            dataset.append((img_path, pid, camid))
            pid_container.add(pid)
        num_pids = len(pid_container)
        # check if pid starts from 0 and increments with 1
        for idx, pid in enumerate(pid_container):
            assert idx == pid, "See code comment for explanation"
        return dataset
```

**Comment by sovrasov** [comment]
Confirm that problem. I also didn't manage to get the SOTA result on MSMT.

**Comment by Cosmopal** [comment]
See issue #23 
Maybe this helps?
